frontend/lib/resolution/maybe-const.cpp

```diff
@@ -145,54 +179,148 @@ void AdjustMaybeRefs::process(const uast::AstNode* symbol,
+  } else if (auto call = ast->toCall()) {
+    auto parent = parsing::parentAst(context, call);
+
+    // Set something general in case we can't find out more details.
+    lhs = { parent, rv.byAst(parent).type() };
+    rhs = { call, rv.byAst(call).type() };
+
+    // If the parent is a call, try to retrieve the formal using the actual.
+    if (auto parentCall = parent->toCall()) {
+      if (call == parentCall->calledExpression()) {
+        CHPL_UNIMPL("Const checking associated actions when call "
+                    "is base expression");
+        return false;
+      } else {
+        for (int i = 0; i < parentCall->numActuals(); i++) {
+          if (call != parentCall->actual(i)) continue;
+
+          auto& reParentCall = rv.byAst(parentCall);
+          auto& mfs = reParentCall.mostSpecific().only();
+          if (!mfs) return false;
+
+          // Try to use more specific details for the LHS.
+          auto& fam = mfs.formalActualMap();
+          if (auto fa = fam.byActualIdx(i)) {
+            auto& formalType = fa->formalType();
+            lhs = { fa->formal(), formalType, mfs.fn(), i };
+          }
+        }
+      }
+    }
+    return true;
   }
 
-  return VALUE; // including IN at least
+  return false;
 }
 
-AdjustMaybeRefs::Access AdjustMaybeRefs::currentAccess() {
-  Access access = VALUE;
-  if (exprStack.size() > 0) {
-    access = exprStack.back().access;
+static const char* initOrAssignStr(AssociatedAction::Action a) {
+  switch (a) {
+    case AssociatedAction::ASSIGN: return "assign";
+    case AssociatedAction::COPY_INIT: return "copy initialize";
+    case AssociatedAction::INIT_OTHER: return "copy initialize";
+    default: break;
   }
-  return access;
+  return nullptr;
 }
 
 void AdjustMaybeRefs::constCheckAssociatedActions(const AstNode* ast, RV& rv) {
-  //ResolvedExpression& re = rv.byAst(ast);
+  auto& re = rv.byAst(ast);
+
+  for (auto& a : re.associatedActions()) {
+    switch (a.action()) {
+      case AssociatedAction::ASSIGN:
+      case AssociatedAction::COPY_INIT:
+      case AssociatedAction::INIT_OTHER: {
+        ExprStackEntry lhs;
+        ExprStackEntry rhs;
+
+        auto fn = a.fn();
+        bool found = findInitOrAssignmentParts(ast, rv, a, lhs, rhs);
+        if (!found) continue;
+
+        Access rhsRequiredAccess;
+        if (a.action() == AssociatedAction::ASSIGN) {
+          CHPL_ASSERT((fn->isMethod() && fn->numFormals() == 3) ||
+                      (fn->numFormals() == 2));
+          rhsRequiredAccess = fn->isMethod() ? fn->formalType(2)
+                                             : fn->formalType(1);
+        } else {
+          CHPL_ASSERT(fn->isMethod() && fn->numFormals() == 2);
+          rhsRequiredAccess = fn->formalType(1);
+        }
 
-  /*
-  TODO:
-  How do I know if it's assigning from or assigning to?
-   bar(foo())
-  bar could have in intent -> assign to formal from foo()
-  bar could have out intent -> assign from formal to foo()
```

Comment:
What was the solution to this conundrum in the end? Are we leaving it as future work or did you finesse the issue by assuming it's whatever would not have been a const-ness error?

Reply:
As far as `const` checking goes, it seems like if `foo()` returns by `const ref` then it cannot be passed to `bar` in the first place. I get a `const` checking error earlier in the pass before the associated action is even checked.

I'm blanking here - is there another pattern I could test here to make sure `out` is doing the right thing?

Reply:
I don't think further action is needed here right now & I'm happy with your answer to my question.